### ISSUE TITLE
Datadog - Profiling the SMS Path

### ISSUE THREAD

**Issue by ldraney** [department-of-veterans-affairs/notification-api#1122]
## Value Statement

**As** VA Notify
**I want to** have accurate representation of our method calls for SMS in each environment
**So that** there is a clear performance indicators per function in our application for SMS

## Additional Info and Resources

Datadog experts and our primary POC recommended that we dedicate time to follow the [Unified Service Tagging](https://docs.datadoghq.com/getting_started/tagging/unified_service_tagging/?tab=ecs#overview), as using the tags is the "secret sauce of how Datadog is working behind the scenes" - Datadog expert.  
They also were surprised that our profiling page does not yet show the sms and email endpoints of our application, and strongly suggested we look into [custom instrumentation](https://docs.datadoghq.com/tracing/trace_collection/custom_instrumentation/#pagetitle), as well as look into our [debug logs](https://docs.datadoghq.com/tracing/troubleshooting/tracer_debug_logs/#enable-debug-mode) to ensure automatic instrumentation is working as it should.  

## Engineering Checklist

- [x] Get the sms endpoint to appear in the profiler page of our application, either by correcting the tagging OR by implementing custom instrumentation in the relevant parts of our application related to the sms path. 
- [x] Preferable: SMS profiler shows all method calls, not just the top level

## Acceptance Criteria

sms path appears here:
![image.png](https://images.zenhubusercontent.com/6361638d438b9c8e844e2b23/a1f06814-0b2a-41f9-9ac4-5ffb62587d92)

## Assumptions

- The Profiler has been set up correctly and is showing up in the dashboard, but likely with incomplete information
- Right now the profiler is blank because Datadog changes were removed due to potential Gunicorn conflict, but that will be resolved once the Datadog is running locally with Docker Compose.

**Comment by ldraney** [comment]
@tabinda-syed   Mel asked the same question, to which I responded:
"The initial profiling configuration isn't difficult, and once anything shows up on the profiling page, then that part is complete.
However, the data that shows up may not actually reflect our application, thus I can't do the test Kyle is talking about.  That is likely due to the tags I need to set up in the Environment Tracing and Logging Ticket"

Mel suggested the following (to which I agreed):
How about this, complete the profiling ticket itself and review it with Kyle. Then determine additional work/tickets as next steps rather than adding more to that ticket.
Since it's new, it's ok if we need to re-assess what's achievable in the first ticket as you learn more in implementation. As you hit bumps in the road just raise them and Kyle can help you determine which are worth digging into more vs what is out of scope.

**Comment by ldraney** [comment]
The engineering checklist may change due to my improving knowledge base.  I'm currently taking the courses offered on [The Datadog Learning Center](https://learn.datadoghq.com/), and in the [dd-fundamentals learning path](https://learn.datadoghq.com/bundles/dd-fundamentals) Datadog 101: Developer course, they add the following snippet to their Docker Compose file:

**Comment by mjones-oddball** [comment]
@cris-oddball for your review as well

**Comment by mjones-oddball** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/VA-Notify-620d21369d810a00146ed9c8/issues/department-of-veterans-affairs/notification-api/1122?planning-poker) with Zenhub @cris-oddball @EvanParish @ianperera @jakehova @justaskdavidb2 @k-macmillan @kalbfled @ldraney

**Comment by ldraney** [comment]
The latest email from the Datadog team after I sent them a detailed report on why our profiler might not be working as expected:

Hey Lucas, 

Thanks for sharing all that (very detailed)!
 
I am consulting with the profiling team now on this. They are still reviewing, but would like the following info:
Operating system type and version (for example, Linux Ubuntu 20.04)
Runtime type, version, and vendor (for example, Python 3.9.5)
tracer debug logs: https://docs.datadoghq.com/tracing/troubleshooting/tracer_debug_logs/#enable-debug-mode
 
Best,
Danny

**Comment by tabinda-syed** [comment]
Adding a "Changes Requested" as Lucas is following up about the local env/AWS checklist item with Kyle.

**Comment by tabinda-syed** [comment]
Removing "Changes Requested." The local env step was removed, and implementation will just involve AWS.

**Comment by mjones-oddball** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/VA-Notify-620d21369d810a00146ed9c8/issues/department-of-veterans-affairs/notification-api/1122?planning-poker) with Zenhub @cris-oddball @EvanParish @ianperera @jakehova @justaskdavidb2 @k-macmillan @kalbfled @ldraney

**Comment by ldraney** [comment]
I got a local environment working, which will be great for testing things. 
You can see that "local" now appears an an env option, and the visible traces are particular to this environment. 
The profiler is still empty, which shouldn't be the case because I've enabled it... but maybe the ddog version is still out of date. 
Then, this may still be a docker-compose issue, so I'll try removing the datadog.yaml file and using only docker-compose, or visa-versa.

**Comment by ldraney** [comment]
We are getting profiler stuff! 
![image.png](https://images.zenhubusercontent.com/6361638d438b9c8e844e2b23/cf851308-3d14-4490-9907-49104e346585)

**Comment by ldraney** [comment]
things seem to be working locally with some adjustments, going to push some task definitions soon for devs app and celery

**Comment by ldraney** [comment]
I think my profiling branch isn't working anymore since I merged the latest updates to my branch.
ddtrace gives me an error:
```
app_1            |   File "/usr/local/lib/python3.8/site-packages/ddtrace/vendor/wrapt/wrappers.py", line 804, in lookup_attribute
app_1            |     return getattr(parent, attribute)
app_1            | AttributeError: type object 'Flask' has no attribute 'before_first_request'
```
Another engineer is also experiencing issues on build, so it may be that my changes won't work until we fix a potential dependency issue.

**Comment by ldraney** [comment]
Okay, so the above was resolved with some work arounds, and again the profiler is successfully working locally. 

But, deploying, even though the env variables are the same as the docker-compose environment, it is still not sending to the profiler. 

I emailed ddog support team, and there's a few things I can try once I can deploy to Dev again.  
Fargate-ecs is just running docker in the background, so I'm not sure why it's not transmitting profiler data.  Is it even sending tracing data?  I'll check that as well.

**Comment by ldraney** [comment]
changing the:
```
tracer.configure(
    hostname='localhost',
    port=8126,
)
```
to localhost rather than datadog-agent got some gunicorn data in the profiler (we've seen this before):
![image.png](https://images.zenhubusercontent.com/6361638d438b9c8e844e2b23/a5dfc6e9-f72c-4322-9279-c7f33f89c2ba)

But, this is not even close to what we are seeing in the docker container.  I'm not sure why. I could be:
1. I haven't modified the task definitions for celery and celery-beat
2. I need to try adding the decorator again

**Comment by ldraney** [comment]
again, the problem is this:
![image.png](https://images.zenhubusercontent.com/6361638d438b9c8e844e2b23/8141d778-b9b2-400a-8164-d3760f69c7c1)
the FLASK is only for docker-compose, and Gunicorn doesn't show up at all. 
But in ECS, on Gunicorn shows up, and not flask. 
Flask looks like this, btw (which seems like much more useful data:)
![image.png](https://images.zenhubusercontent.com/6361638d438b9c8e844e2b23/7b66e5c8-54b9-45dc-b0a3-6af123e266d5)

**Comment by ldraney** [comment]
It should be noted that the AWS logs give me:
```
WARNING  /usr/local/lib/python3.8/site-packages/ddtrace/profiling/scheduler.py:55: Unable to export profile: ddtrace.profiling.exporter.http.UploadFailed: [Errno -2] No address found. Ignoring.
```

**Comment by ldraney** [comment]
well the answer is local dockerfile starts with:

and deployed dockerfile starts with:
```
CMD ["./scripts/run_app.sh"]
```
which runs:
```
flask --debug run -p 6011 --host=0.0.0.0
```
```
CMD ["sh", "-c", "gunicorn -c gunicorn_config.py application"]
```

**Comment by ldraney** [comment]
These outputs are actually similar, though docker seems to pick up other functions (typing, posixpath) that are perhaps not being run right off the bat in ECS
![image.png](https://images.zenhubusercontent.com/6361638d438b9c8e844e2b23/b234a937-ca75-4e4e-858f-fde967aa94ba)

But, when I send an SMS in Postman, it doesn't show up in the profiler.

**Comment by ldraney** [comment]
But I am noticeably getting tracing data from dev for Flask:
![image.png](https://images.zenhubusercontent.com/6361638d438b9c8e844e2b23/9ef6f5af-3d66-4ac8-a65d-60cfb276d189)
So, why is it not in the profiler?

**Comment by ldraney** [comment]
SWEET!  Finally getting some more info on the gunicorn start! 
In the last commit:
1. updated celery and celery beat's task definitions for the Datadog container to have some additional env variables. 
2. adding ddtrace to the gunicorn command in our Dockerfile

**Comment by ldraney** [comment]
I sent this beautiful message to the ddog team:
I'm getting more data, but in short, Datadog Profiler does not recognize our methods at all.It's all useless gunicorn and flask data.
Task definition:
      "name": "datadog-agent",
      "image": "171875617347.dkr.ecr.us-gov-west-1.amazonaws.com/datadog/agent:latest",
      "essential": true,
      "logConfiguration": {
        "logDriver": "awslogs",
        "options": {
          "awslogs-group": "dev-notification-api-datadog-log-group",
          "awslogs-region": "us-gov-west-1",
          "awslogs-stream-prefix": "ecs"
        }
      },
      "portMappings": [
        {
          "containerPort": 8125,
          "hostPort": 8125,
          "protocol": "udp"
        },
        {
          "containerPort": 8126,
          "hostPort": 8126,
          "protocol": "tcp"
        }
      ],
      "environment": [
        {
          "name": "ECS_FARGATE",
          "value": "true"
        },
        {
          "name": "DD_SITE",
          "value": "ddog-gov.com"
        },
        {
          "name": "DD_APM_ENABLED",
          "value": "true"
        },
        {
          "name": "DD_APM_NON_LOCAL_TRAFFIC",
          "value": "true"
        },
        {
          "name": "DD_PROCESS_AGENT_ENABLED",
          "value": "true"
        },
        {
          "name": "DD_ENV",
          "value": "dev"
        },
        {
          "name": "DD_PROFILING_MAX_EVENTS",
          "value": "10000"
        }
      ],
      "secrets": [
        {
          "name": "DD_API_KEY",
          "valueFrom": "....."
        }

Our dockerfile:
CMD ["sh", "-c", "ddtrace-run gunicorn -c gunicorn_config.py application"]

Our application.py
#!/usr/bin/env python
from __future__ import print_function
import os

import sentry_sdk
# from ddtrace import patch_all
from ddtrace import config, patch_all, tracer
from ddtrace.profiling import Profiler
from flask import Flask
from sentry_sdk.integrations.flask import FlaskIntegration
from werkzeug.middleware.proxy_fix import ProxyFix

from app import create_app

from dotenv import load_dotenv

load_dotenv()

sentry_sdk.init(
    dsn=os.environ.get("SENTRY_URL", ""),
    integrations=[FlaskIntegration()],
    release=os.environ.get("GIT_SHA", ""),
)

application = Flask("app")
application.wsgi_app = ProxyFix(application.wsgi_app)
create_app(application)

tracer.configure(
    hostname='localhost',
    port=8126,
)

# this starts the ddtrace tracer and configures it to the right port and URL
patch_all()

config.profiling.enabled = True
profiler = Profiler()
profiler.start()

And a screenshot for our profiler as of Apr 27, 4:08 pm – Apr 27, 8:08 pm (EST):
 

But Tracing seems to be working fine (we are trying to, most importantly, trace our SMS path.  We are trying to profile our SMS path, at least):
 

As a final note, our cloudwatch logs say:
ERROR  /usr/local/lib/python3.8/site-packages/ddtrace/profiling/profiler.py:270: Failed to start collector MemoryCollector(status=<ServiceStatus.STOPPED: 'stopped'>, recorder=Recorder(default_max_events=16384, max_events={<class 'ddtrace.profiling.collector.stack_event.StackSampleEvent'>: 30000, <class 'ddtrace.profiling.collector.stack_event.StackExceptionSampleEvent'>: 15000, <class 'ddtrace.profiling.collector.memalloc.MemoryAllocSampleEvent'>: 1920, <class 'ddtrace.profiling.collector.memalloc.MemoryHeapSampleEvent'>: None}), _max_events=16, max_nframe=64, heap_sample_size=1048576, ignore_profiler=False), disabling.

We have tried to resolve the above error by adding this to the task definitions:
        {
          "name": "DD_PROFILING_MAX_EVENTS",
          "value": "10000"
        }

and increasing the memory on the tasks as well up to 4gb.  Maybe this is related?  

Thanks for any insight on this issue!

**Comment by ldraney** [comment]
As far as the profiler memory error is concerned:
looks like that might be fixed if we apply the latest patch [v1.12.1](https://ddtrace.readthedocs.io/en/stable/_release_notes_all.html)

**Comment by ldraney** [comment]
I've confirmed that updated to ddtrace 1.12.1 fixed the profiler memory error, thought I did increase the memory on the tasks, so that may be a factor as well.  
But, still no useful data in the profiler.   I'm now pinging the engineering team and doing research and using resources the best I can.

**Comment by ldraney** [comment]
Kimball gave me good advice in that the [ddtrace documentation on proflier](https://ddtrace.readthedocs.io/en/stable/basic_usage.html#via-module) indicated that profile.start() might be working against me, and would explain the limitations we see.  

However, when I remove this part, nothing shows up in our dashboard at all.  But I do fell this is the right direction.

**Comment by ldraney** [comment]
just by importing the profiling.auto module, I'm getting more data: 
![image.png](https://images.zenhubusercontent.com/6361638d438b9c8e844e2b23/23078532-0afe-4f8c-9715-c1701413a859)
let's see how this appears in ECS

**Comment by ldraney** [comment]
I played with a few more things, but I couldn't get the profiler to start on its own. 

I still have to start it with:
```
config.profiling.enabled = True
profiler = Profiler()
profiler.start()
```
But, it does seem to give new information.  For example run_migrations now shows up when running locally:
![image.png](https://images.zenhubusercontent.com/6361638d438b9c8e844e2b23/af83714c-8e63-479c-8b94-7e50ca3943fe)

**Comment by ldraney** [comment]
I wrote a message to Brent with the VA ddog team with similar details as above, we'll see what he says. 

There are various env variables I will be playing with on the [ddtrace documentation](https://ddtrace.readthedocs.io/en/stable/configuration.html#DD_TRACE_DEBUG), hopefully that reveals some new information.

**Comment by ldraney** [comment]
BTW, as a disclaimer, the [Datadog Docs](https://docs.datadoghq.com/profiler/enabling/python/#caveats) do indicate using the ddtract.profiling rather than even mentioning ddtrace.profiling.auto
but the advice Kimball gave me from the ddtrace docs seems more accurate to the situation.

**Comment by ldraney** [comment]
I noticed this in [the file Kimball sent to me](https://github.com/DataDog/dd-trace-py/blob/1.x/ddtrace/bootstrap/sitecustomize.py):
```
    if profiling:
        log.debug("profiler enabled via environment variable")
        import ddtrace.profiling.auto  # noqa: F401
```
I certainly have profiler enabled everywhere I could figure out, so I tried to see if there was some kind of redundancy conflict by removing the profiler code from python.

**Comment by ldraney** [comment]
I seem to be getting some application data:
![image.png](https://images.zenhubusercontent.com/6361638d438b9c8e844e2b23/f8507cb6-4d49-4477-884e-e4959e875247)

**Comment by ldraney** [comment]
Look at the highlights, they are all parts of our application:
![image.png](https://images.zenhubusercontent.com/6361638d438b9c8e844e2b23/f2fadf54-63ae-406b-8e4d-7e492a8c76cf)
I will try deploying again soon.

**Comment by ldraney** [comment]
BTW, `/app/app/notifications/process_notifications.py ` is where `deliver_sms` and `deliver_sms_with_rate_limiting` are found.

**Comment by ldraney** [comment]
I'm changing this in our application.py:
```
tracer.configure(
    hostname='datadog-agent',
    port=8126,
)
```
to this in our docker-compose:
```
    environment:
      - DATADOG_AGENT_HOSTNAME=datadog-agent
      - DATADOG_AGENT_PORT=8126
```
and this in our task definition for the notification-api image:
```
        {
          "name": "DATADOG_AGENT_HOSTNAME",
          "value": "localhost"
        },
        {
          "name": "DATADOG_AGENT_PORT",
          "value": "8126"
        },
```

**Comment by ldraney** [comment]
The above changes should have been simple, but things are very inconsistent.  
If I updated here every time my local configuration didn't show up in the profiler when I hadn't made any file changes (on the same branch), it would be an appropriate disclaimer for how long this is taking me. 
I may actually start doing that if I continue to face these problems.

**Comment by ldraney** [comment]
I've been beating my head against some docker inconsistencies with Datadog nuances.
So, I won't include my docker-compose changes as a part of this branch.  In other words, I've realized have the local datadog environment is out of scope of this ticket; which is obvious, except that it was a huge help getting in helping me get to this point.
But I'll take those parts out, do a few more deploys to make sure things are clean and working in ECS, then open the PR.

**Comment by ldraney** [comment]
Just read this on the [datadog documentation](https://docs.datadoghq.com/tracing/guide/setting_primary_tags_to_scope/?tab=environmentvariables):
If you configure the Agent with environment variables, as is common with Docker and ECS installations, pass the following environment variable to the trace Agent after upgrading the Docker image.
DD_APM_FEATURES=enable_cid_stats

**Comment by ldraney** [comment]
I'm finally getting the profiler to work how I want:

But I have some potential redundancy, because I include the following env variables for both the notification-api and the datadog agent containers:
In your case, both the notification-api and datadog-agent containers share some common environment variables:

ECS_FARGATE
DD_SITE
DD_APM_ENABLED
DD_PROCESS_AGENT_ENABLED
DD_ENV
DD_SERVICE
DD_APM_FEATURES
DD_PROFILING_ENABLE_CODE_PROVENANCE

I likely need these environment variables in the notification-api container to ensure that the Datadog agent is properly set up for monitoring and profiling the application. On the other hand, the datadog-agent container needs these environment variables for its own configuration.  That is the result of my experience, and is something I'd like to move forward with based on the time it takes for me to test any changes in ECS. 

I can keep the redundancy if I think that both containers require these shared environment variables. However, if I find that some of them are not necessary for one of the containers, I can remove them from the respective container definition.  

I'll keep playing a little though, to ensure this is the case. 

Here is how the profiler looks now (and seems good to me):

![image.png](https://images.zenhubusercontent.com/6361638d438b9c8e844e2b23/b807543b-463b-4303-9eae-9bf502849d6d)

**Comment by ldraney** [comment]
TLDR for previous comment:
I would like to make a ticket for "reduce potential redundancy for Datadog in Task Definitions", because getting this will take some time. 

This also answers the question of "How did I get this working?", I think it's because of these env variables being available to both containers that things started working properly.

**Comment by ldraney** [comment]
I decided to get more scientific with how these variables are performing. 
I've been playing with task definitions in the AWS console for the dev environment:

# markdown test

**Comment by ldraney** [comment]
TLDR the previous comment: I've realized that [ddtrace](https://ddtrace.readthedocs.io/en/stable/configuration.html) and the [Agent](https://docs.datadoghq.com/containers/docker/?tab=standard#environment-variables) have similar env variables, and both need to be declared in our situation.  So its important that there is consistency, and that all necessary variables are declared.

**Comment by ldraney** [comment]
Here's a summary of what I currently have set as environment variables related to Datadog for each container:
notification-api container (mentioned in the ddtrace documentation):
DD_ENV: "dev"
DD_SITE: "ddog-gov.com"
DD_PROFILING_ENABLED: "True"
DD_SERVICE: "notification-api"
DD_PROFILING_ENABLE_CODE_PROVENANCE: "True"

not mentioned in ddtrace documentation:
DD_APM_NON_LOCAL_TRAFFIC: "true"
DD_LOGS_ENABLED: "true"
DD_PROCESS_AGENT_ENABLED: "true"
DD_APM_FEATURES: "enable_cid_stats"
DD_APM_ENABLED: "true"

datadog-agent container definitions all mentioned in the documentation:
DD_APM_NON_LOCAL_TRAFFIC: "true"
DD_LOGS_ENABLED: "true"
DD_PROCESS_AGENT_ENABLED: "true"
ECS_FARGATE: "true"
DD_SITE: "ddog-gov.com"
DD_APM_ENABLED: "true"
DD_ENV: "dev"
DD_SERVICE: "notification-api"
DD_APM_FEATURES: "enable_cid_stats"
DD_PROFILING_ENABLE_CODE_PROVENANCE: "true"

**Comment by ldraney** [comment]
also I believe the Agent likes lowercase `true`, whereas [ddtrace ](https://ddtrace.readthedocs.io/en/stable/configuration.html#DD_PROFILING_ENABLED) likes the upper case `True`
